src/Microsoft.ML.Transforms/CategoricalTransform.cs

```diff
@@ -152,88 +136,115 @@ public static IDataTransform Create(IHostEnvironment env, Arguments args, IDataV
             h.CheckValue(args, nameof(args));
             h.CheckValue(input, nameof(input));
             h.CheckUserArg(Utils.Size(args.Column) > 0, nameof(args.Column));
-            return CreateTransformCore(
-                args.OutputKind,
-                args.Column,
-                args.Column.Select(col => col.OutputKind).ToList(),
-                 TermTransform.Create(h, args, args.Column, input),
-                h,
-                env);
+
+            var columns = new List<CategoricalEstimator.ColumnInfo>();
+            foreach (var column in args.Column)
+            {
+                var col = new CategoricalEstimator.ColumnInfo(
+                    column.Source ?? column.Name,
+                    column.Name,
+                    column.OutputKind ?? args.OutputKind,
+                    column.MaxNumTerms ?? args.MaxNumTerms,
+                    column.Sort ?? args.Sort,
+                    column.Term ?? args.Term);
+                col.SetTerms(column.Terms);
+                columns.Add(col);
+            }
+            return new CategoricalEstimator(env, columns.ToArray()).Fit(input).Transform(input) as IDataTransform;
         }
 
-        public static IDataTransform CreateTransformCore(
-            OutputKind argsOutputKind,
-            OneToOneColumn[] columns,
-            List<OutputKind?> columnOutputKinds,
-            IDataTransform input,
-            IHost h,
-            IHostEnvironment env,
-            CategoricalHashTransform.Arguments catHashArgs = null)
+    }
+
+    public sealed class CategoricalEstimator : IEstimator<ITransformer>
```

Comment:
>IEstimator<ITransformer> [](start = 47, length = 24)

Should not the categorical estimator have a more specific type? We have something that we want to publish, do we not? We want something that is (ultimately) capable of exposing whatever `ToKey` mapping was learnt in the first step. This strikes me as the only real reason why we would actually have the complication of an implemented type in the first palce.

Reply:
Yes thanks.

---
In reply to: [218176602](https://github.com/dotnet/machinelearning/pull/899#discussion_r218176602) [](ancestors = 218176602,217570810)